src/include/common/concurrent_bitmap.h

```diff
@@ -88,6 +88,78 @@ class RawConcurrentBitmap {
     return false;
   }
 
+  /**
+   * Returns the position of the first unset bit, if it exists.
+   * Note that this result is immediately stale.
+   * @param num_bits number of bits in the bitmap.
+   * @param start_pos start searching from this bit location.
+   * @param[out] out_pos the position of the first unset bit will be written here, if it exists.
+   * @return true if an unset bit was found, and false otherwise.
+   */
+  bool FirstUnsetPos(uint32_t num_bits, uint32_t start_pos, uint32_t *out_pos) {
+    uint32_t num_bytes = BitmapSize(num_bits);  // maximum number of bytes in the bitmap
+    uint32_t byte_pos = start_pos / BYTE_SIZE;  // current byte position
+    uint32_t search_width;                      // number of bits searched at a time
+
+    // optimization: don't search wider than the number of bits we have
+    if (num_bits <= 8) {
+      search_width = 8;
+    } else if (num_bits <= 16) {
+      search_width = 16;
+    } else if (num_bits <= 32) {
+      search_width = 32;
+    } else {
+      search_width = 64;
+    }
+
+    while (byte_pos < num_bytes) {
+      // each case loads a word of search_width bits and casts it as a signed type.
+      // if all bits are set (equals -1), then we can refine our search.
+      switch (search_width) {
+        default:
+        case 64:
+          if (static_cast<std::atomic<int64_t>>(bits_[byte_pos]).load() != -1) {
```

Comment:
You can probably templatize this on T, considering that earlier you would do sizeof(T), and otherwise these branches look identical.